cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ResourceDistributionGoal.java

```diff
@@ -462,11 +727,139 @@ private boolean isLoadUnderBalanceUpperLimitAfterChange(ClusterModel clusterMode
+    }
+
+    if (!isHostUnderUpperLimit) {
+      return false;
+    }
+    // 2. Host above balance lower limit check.
+    boolean isHostAboveLowerLimit;
+    if (sourceUtilizationDelta < 0) {
+      double sourceHostBalanceLowerLimit = sourceReplica.broker().host().capacityFor(resource()) * _balanceLowerThreshold;
+      isHostAboveLowerLimit = sourceHostUtilization + sourceUtilizationDelta >= sourceHostBalanceLowerLimit;
+    } else {
+      double destinationHostBalanceLowerLimit = destinationReplica.broker().host().capacityFor(resource()) * _balanceLowerThreshold;
+      isHostAboveLowerLimit = destinationHostBalanceLowerLimit - sourceUtilizationDelta >= destinationHostBalanceLowerLimit;
+    }
+
+    return isHostAboveLowerLimit;
+  }
+
+  private boolean isBrokerLoadStillWithinLimitsAfterSwap(double sourceUtilizationDelta,
+                                                         Replica sourceReplica,
+                                                         Replica destinationReplica) {
+    double sourceBrokerUtilization = sourceReplica.broker().load().expectedUtilizationFor(resource());
+    double destinationBrokerUtilization = destinationReplica.broker().load().expectedUtilizationFor(resource());
+
+    // 1. Broker under balance upper limit check.
+    boolean isBrokerUnderUpperLimit;
+    if (sourceUtilizationDelta > 0) {
+      double sourceBrokerBalanceUpperLimit = sourceReplica.broker().capacityFor(resource()) * _balanceUpperThreshold;
+      isBrokerUnderUpperLimit = sourceBrokerUtilization + sourceUtilizationDelta <= sourceBrokerBalanceUpperLimit;
+    } else {
+      double destinationBrokerBalanceUpperLimit = destinationReplica.broker().capacityFor(resource()) * _balanceUpperThreshold;
+      isBrokerUnderUpperLimit = destinationBrokerUtilization - sourceUtilizationDelta <= destinationBrokerBalanceUpperLimit;
+    }
+
+    if (!isBrokerUnderUpperLimit) {
+      return false;
+    }
+
+    // 2. Broker above balance lower limit check.
+    boolean isBrokerAboveLowerLimit;
+    if (sourceUtilizationDelta < 0) {
+      double sourceBrokerBalanceLowerLimit = sourceReplica.broker().capacityFor(resource()) * _balanceLowerThreshold;
+      isBrokerAboveLowerLimit = sourceBrokerUtilization + sourceUtilizationDelta >= sourceBrokerBalanceLowerLimit;
+    } else {
+      double destinationBrokerBalanceLowerLimit = destinationReplica.broker().capacityFor(resource()) * _balanceLowerThreshold;
+      isBrokerAboveLowerLimit = destinationBrokerUtilization - sourceUtilizationDelta >= destinationBrokerBalanceLowerLimit;
+    }
+
+    return isBrokerAboveLowerLimit;
+  }
+
   /**
    * @param clusterModel the cluster topology and load.
    * @return the utilization upper threshold in percent for the {@link #resource()}
    */
-  protected double balanceUpperThreshold(ClusterModel clusterModel) {
+  private double balanceUpperThreshold(ClusterModel clusterModel) {
```

Comment:
Any reason to make these changes?

Reply:
Added two other protected getters -- without computation.